collector/prometheus.go

```diff
@@ -224,22 +231,37 @@ func (c *PrometheusCollector) getQueryRangeResults(queries *querys, results *map
 		if !ok {
 			return fmt.Errorf("expected a matrix in response to query, got a %v", queryResult.Type())
 		}
-
 		results.iterateMatrix(matrix, query)
 	}
+
+	if len(queriesToRetry) > 0 {
+		retries--
+		waitTime := time.Duration(5-retries) * time.Second
```

Comment:
You could use exponential backoff here rather than a linear algorithm. If there is potential for a thundering herd, adding some jitter to the wait time will mitigate that.

Reply:
How do we determine the risk of thundering herd? I can implement an exponential backoff, but because I don't understand the risk, I don't understand the benefit. I only see a downside because unless I change some testing logic, this is going to make our tests take a long time.

Reply:
> How do we determine the risk of thundering herd?

Consider the number of copies of this code running and if there are potential triggers that could cause them all to run synchronously, such as a process restart or time based run. It may not be an issue in the current architecture, but it's just something to consider when implementing retry code.

Adding some random fractional jitter to the sleep time helps break up the requests so they don't all arrive at the same time even if the processes were all triggered at the same time.

> I can implement an exponential backoff, but because I don't understand the risk, I don't understand the benefit

Exponential backoff can help with two things: 1) Reducing overall wait time and 2) Reducing the number of requests made to a resource.

If it doesn't accomplish those two things then it doesn't add any benefit over a linear retry timer.

For example, say the wait time is 0.5, then 1, then 2 and it succeeded on the third retry. That's less wait time than five seconds per retry.

Reply:
The risk here is pretty low then. Only a single operator runs on a cluster, and there is only a single CR created that the operator handles. Everything runs synchronously.

While the retry time was not exponential, it wasn't just 5 seconds between retries; it was 1, 2, 3, 4, 5 seconds.

Reply:
Sounds good. Sorry I misread the sleep times.